confgenerator/add_lua_script.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package confgenerator
+
+import (
+  "fmt"
+  "io/ioutil"
+  "os"
+  "path/filepath"
+)
+
+const addLogNameLuaScriptContents string = `
+function split (inputstr, sep)
+  if sep == nil then
+    sep = "%s"
+  end
+  local t={}
+  for str in string.gmatch(inputstr, "([^"..sep.."]+)") do
+    table.insert(t, str)
+  end
+  return t
+end
+
+function add_log_name(tag, timestamp, record)
+  -- Split the tag by . to seperate the pipeline_id from the rest
+  local split_tag = split(tag, '.')
+  
+  -- Join the rest of the tag without the pipeline_id
+  local pipeline_id = table.remove(split_tag, 1)
+  local log_name = table.concat(split_tag, ".")
+  
+  -- Replace the record with one with the log name
+  record["logging.googleapis.com/logName"] = log_name
+  
+  -- Use code 2 to replace the original record but keep the original timestamp
+  return 2, timestamp, record
+end
+`
+
+// writeForwardScript writes the above Lua script to the given path so it can be
+// used by the logging subagent.
+//
+// TODO(ridwanmsharif): Replace this with in-config script when
+//   fluent/fluent-bit#4634 is supported.
+func writeForwardScript(path string) error {
```

Comment:
Per sync offline, we need more generic util for LUA script (needed by `modify_fields` as well). Will defer to @quentinmit and @ridwanmsharif to align on the desired file location and method name. I'm OK with a follow up PR as long as we are committed to getting it refactored in the near term.

Reply:
We haven't decided yet but this will be done in a follow-up PR. The config writing code will be folded in with this as well, and we'll have a common util that will write write files to disk (configs, Lua scripts, anything else we will need in the future). Keeping it separate from this PR so this PR doesn't get too unwieldy